File: lib/features/simulation/domain/simulation_calculator.dart
import 'dart:math';

import 'package:savvy/core/constants/financial_enums.dart';
import 'package:savvy/core/utils/financial_calculator.dart';
import 'package:savvy/features/dashboard/domain/models/month_summary.dart';
import 'package:savvy/features/simulation/domain/models/simulation_change.dart';
import 'package:savvy/features/simulation/domain/models/simulation_result.dart';

/// BL-001: All simulation calculations happen here — nowhere else.
class SimulationCalculator {
  // ─── Main Entry Point ────────────────────────────────────────────

  /// Calculate a full scenario from composable changes against current budget.
  ///
  /// [baseItems] — when provided, the 12-month projection uses a dynamic base
  /// (each month filters active recurring items by their start/end dates).
  /// When empty, falls back to the static [currentBudget] snapshot (old behaviour).
  static SimulationResult calculateScenario({
    required List<SimulationChange> changes,
    required MonthSummary currentBudget,
    List<MonthLineItem> existingIncomeItems = const [],
    List<MonthLineItem> existingExpenseItems = const [],
    List<ProjectionBaseItem> baseItems = const [],
  }) {
    if (changes.isEmpty) {
      return _emptyResult(currentBudget, existingIncomeItems, existingExpenseItems, baseItems);
    }

    double totalIncomeImpact = 0;
    double totalExpenseImpact = 0;
    final changeResults = <ChangeResult>[];

    for (final change in changes) {
      final result = _calculateChange(change, currentBudget);
      changeResults.add(result);

      if (result.monthlyImpact > 0) {
        totalIncomeImpact += result.monthlyImpact;
      } else {
        totalExpenseImpact += result.monthlyImpact.abs();
      }
    }

    final newIncome = currentBudget.totalIncome + totalIncomeImpact;
    final newExpense = currentBudget.totalExpense + totalExpenseImpact;
    final newNet = newIncome - newExpense;
    final monthlyNetImpact = totalIncomeImpact - totalExpenseImpact;

    final newExpenseRate = newIncome > 0 ? newExpense / newIncome : 0.0;
    final newSavingsRate = newIncome > 0
        ? ((newIncome - newExpense) / newIncome).clamp(0.0, 1.0)
        : 0.0;

    // Affordability based on total new expense-only impacts vs income
    AffordabilityStatus? affordability;
    if (totalExpenseImpact > 0) {
      affordability = FinancialCalculator.loanAffordability(
        monthlyPayment: totalExpenseImpact,
        monthlyIncome: newIncome,
      );
    }

    // 12-month projection
    final projection = _buildProjection(
      changes: changes,
      changeResults: changeResults,
      currentBudget: currentBudget,
      existingIncomeItems: existingIncomeItems,
      existingExpenseItems: existingExpenseItems,
      baseItems: baseItems,
    );

    return SimulationResult(
      currentIncome: currentBudget.totalIncome,
      currentExpense: currentBudget.totalExpense,
      currentNet: currentBudget.netBalance,
      newIncome: newIncome,
      newExpense: newExpense,
      newNet: newNet,
      monthlyNetImpact: monthlyNetImpact,
      annualNetImpact: _calculateAnnualImpact(changes, changeResults),
      newSavingsRate: newSavingsRate,
      newExpenseRate: newExpenseRate,
      affordability: affordability,
      changeResults: changeResults,
      monthlyProjection: projection,
    );
  }

  // ─── Per-Change Calculators ──────────────────────────────────────

  static ChangeResult _calculateChange(
    SimulationChange change,
    MonthSummary budget,
  ) {
    return switch (change) {
      CreditChange() => _calcCredit(change, budget),
      HousingChange() => _calcHousing(change, budget),
      CarChange() => _calcCar(change, budget),
      RentChangeChange() => _calcRentChange(change),
      SalaryChangeChange() => _calcSalaryChange(change),
      IncomeChange() => _calcIncome(change),
      ExpenseChange() => _calcExpense(change),
      InvestmentChange() => _calcInvestment(change),
    };
  }

  static ChangeResult _calcCredit(CreditChange c, MonthSummary budget) {
    return _calcLoanBased(
      change: c,
      loanPrincipal: c.principal,
      annualRate: c.annualRate,
      termMonths: c.termMonths,
      downPayment: 0,
      monthlyExtras: 0,
    );
  }

  static ChangeResult _calcHousing(HousingChange c, MonthSummary budget) {
    return _calcLoanBased(
      change: c,
      loanPrincipal: c.price - c.downPayment,
      annualRate: c.annualRate,
      termMonths: c.termMonths,
      downPayment: c.downPayment,
      monthlyExtras: c.monthlyExtras,
    );
  }

  static ChangeResult _calcCar(CarChange c, MonthSummary budget) {
    return _calcLoanBased(
      change: c,
      loanPrincipal: c.price - c.downPayment,
      annualRate: c.annualRate,
      termMonths: c.termMonths,
      downPayment: c.downPayment,
      monthlyExtras: c.monthlyRunningCosts,
    );
  }

  /// Shared loan calculation for credit, housing, and car changes.
  static ChangeResult _calcLoanBased({
    required SimulationChange change,
    required double loanPrincipal,
    required double annualRate,
    required int termMonths,
    required double downPayment,
    required double monthlyExtras,
  }) {
    if (loanPrincipal <= 0) {
      return ChangeResult(
        change: change,
        monthlyImpact: -monthlyExtras,
        totalCost: downPayment,
      );
    }

    final rate = annualRate / 100;
    final monthly = FinancialCalculator.monthlyLoanPayment(
      principal: loanPrincipal,
      annualRate: rate,
      termMonths: termMonths,
    );
    final total = monthly * termMonths;
    final interest = total - loanPrincipal;
    final totalMonthly = monthly + monthlyExtras;

    return ChangeResult(
      change: change,
      monthlyImpact: -totalMonthly,
      totalCost: total + downPayment + (monthlyExtras * termMonths),
      totalInterest: interest,
      amortizationSchedule: _amortizationSchedule(
        principal: loanPrincipal,
        annualRate: rate,
        termMonths: termMonths,
      ),
    );
  }

  static ChangeResult _calcRentChange(RentChangeChange c) {
    final diff = c.newRent - c.currentRent;
    return ChangeResult(
      change: c,
      monthlyImpact: -diff,
    );
  }

  /// Returns the rent diff for a given projection month index,
  /// applying annual compounding if annualIncreaseRate > 0.
  static double _rentDiffForMonth(RentChangeChange c, int monthIndex) {
    final baseDiff = c.newRent - c.currentRent;
    if (c.annualIncreaseRate <= 0) return baseDiff;
    final yearsPassed = (monthIndex / 12).floor();
    return baseDiff * pow(1 + c.annualIncreaseRate / 100, yearsPassed);
  }

  static ChangeResult _calcSalaryChange(SalaryChangeChange c) {
    final currentNets = FinancialCalculator.resolveAllMonths(
      amount: c.currentGross,
      isGross: true,
    );
    final newNets = FinancialCalculator.resolveAllMonths(
      amount: c.newGross,
      isGross: true,
    );

    final currentAvg = currentNets.reduce((a, b) => a + b) / 12;
    final newAvg = newNets.reduce((a, b) => a + b) / 12;
    final monthlyDelta = newAvg - currentAvg;

    return ChangeResult(
      change: c,
      monthlyImpact: monthlyDelta,
      salaryImpact: SalaryImpact(
        currentGross: c.currentGross,
        newGross: c.newGross,
        currentNetAvg: currentAvg,
        newNetAvg: newAvg,
        monthlyNetDelta: monthlyDelta,
        currentMonthlyNets: currentNets,
        newMonthlyNets: newNets,
      ),
    );
  }

  static ChangeResult _calcIncome(IncomeChange c) {
    return ChangeResult(
      change: c,
      monthlyImpact: c.amount,
    );
  }

  static ChangeResult _calcExpense(ExpenseChange c) {
    return ChangeResult(
      change: c,
      monthlyImpact: -c.amount,
    );
  }

  static ChangeResult _calcInvestment(InvestmentChange c) {
    double totalReturn;
    if (c.isCompound) {
      final monthlyRate = c.annualReturnRate / 100 / 12;
      totalReturn =
          c.principal * (pow(1 + monthlyRate, c.termMonths) - 1).toDouble();
    } else {
      totalReturn =
          c.principal * (c.annualReturnRate / 100) * (c.termMonths / 12);
    }
    final monthlyReturn = totalReturn / c.termMonths;

    return ChangeResult(
      change: c,
      monthlyImpact: monthlyReturn,
      investmentImpact: InvestmentImpact(
        principal: c.principal,
        totalReturn: totalReturn,
        monthlyReturn: monthlyReturn,
        totalValue: c.principal + totalReturn,
        termMonths: c.termMonths,
        isCompound: c.isCompound,
      ),
    );
  }

  // ─── Annual Impact Calculator ────────────────────────────────────

  /// Computes the correct annual net impact, accounting for:
  /// - Salary changes: sum of all 12 months' actual deltas (tax brackets vary)
  /// - Term-limited changes: capped at actual term (not always 12 months)
  static double _calculateAnnualImpact(
    List<SimulationChange> changes,
    List<ChangeResult> changeResults,
  ) {
    double annual = 0;
    for (int i = 0; i < changes.length; i++) {
      final change = changes[i];
      final result = changeResults[i];

      if (change is SalaryChangeChange && result.salaryImpact != null) {
        final si = result.salaryImpact!;
        annual += si.newMonthlyNets.reduce((a, b) => a + b) -
            si.currentMonthlyNets.reduce((a, b) => a + b);
      } else {
        final effectiveMonths = switch (change) {
          CreditChange c => min(c.termMonths, 12),
          HousingChange c => min(c.termMonths, 12),
          CarChange c => min(c.termMonths, 12),
          InvestmentChange c => min(c.termMonths, 12),
          _ => 12,
        };
        annual += result.monthlyImpact * effectiveMonths;
      }
    }
    return annual;
  }

  // ─── 12-Month Projection Builder ────────────────────────────────

  static List<MonthProjection> _buildProjection({
    required List<SimulationChange> changes,
    required List<ChangeResult> changeResults,
    required MonthSummary currentBudget,
    required List<MonthLineItem> existingIncomeItems,
    required List<MonthLineItem> existingExpenseItems,
    List<ProjectionBaseItem> baseItems = const [],
  }) {
    final now = DateTime.now();
    final projections = <MonthProjection>[];
    double cumulative = 0;

    for (int i = 0; i < 12; i++) {
      final projMonth = DateTime(now.year, now.month + i);
      final ym =
          '${projMonth.year}-${projMonth.month.toString().padLeft(2, '0')}';
      final monthLabel = FinancialCalculator.monthShortNamesTR[
          (projMonth.month - 1) % 12];

      // Base: dynamic (recurring items filtered by active date range)
      //       or static fallback (current budget snapshot)
      double monthIncome;
      double monthExpense;
      final incomeItems = <MonthLineItem>[];
      final expenseItems = <MonthLineItem>[];

      if (baseItems.isNotEmpty) {
        monthIncome = 0;
        monthExpense = 0;
        for (final item in baseItems) {
          if (!item.isActiveOn(projMonth)) continue;
          final amount = item.grossAmount != null
              ? FinancialCalculator.resolveNetForMonth(
                  amount: item.grossAmount!,
                  isGross: true,
                  month: projMonth.month,
                )
              : item.netAmount;
          if (item.isIncome) {
            monthIncome += amount;
            incomeItems.add(MonthLineItem(label: item.label, amount: amount));
          } else {
            monthExpense += amount;
            expenseItems.add(MonthLineItem(label: item.label, amount: amount));
          }
        }
      } else {
        monthIncome = currentBudget.totalIncome;
        monthExpense = currentBudget.totalExpense;
        incomeItems.addAll(existingIncomeItems);
        expenseItems.addAll(existingExpenseItems);
      }

      // Apply each change
      for (int ci = 0; ci < changes.length; ci++) {
        final change = changes[ci];
        final result = changeResults[ci];

        // Check term limits for loan-based changes
        final isWithinTerm = switch (change) {
          CreditChange c => i < c.termMonths,
          HousingChange c => i < c.termMonths,
          CarChange c => i < c.termMonths,
          InvestmentChange c => i < c.termMonths,
          _ => true,
        };

        if (!isWithinTerm) continue;

        // Rent changes with annual increase: recompute diff for this month
        if (change is RentChangeChange && change.annualIncreaseRate > 0) {
          final rentDiff = _rentDiffForMonth(change, i);
          monthExpense += rentDiff;
          expenseItems.add(MonthLineItem(
            label: change.label,
            amount: rentDiff,
            isSimulated: true,
          ));
          continue;
        }

        // Salary changes use per-month net (Turkish tax brackets vary)
        if (change is SalaryChangeChange && result.salaryImpact != null) {
          final si = result.salaryImpact!;
          final monthIdx = (projMonth.month - 1) % 12;
          final delta = si.newMonthlyNets[monthIdx] - si.currentMonthlyNets[monthIdx];
          monthIncome += delta;
          incomeItems.add(MonthLineItem(
            label: change.label,
            amount: delta,
            isSimulated: true,
          ));
          continue;
        }

        if (result.monthlyImpact > 0) {
          monthIncome += result.monthlyImpact;
          incomeItems.add(MonthLineItem(
            label: change.label,
            amount: result.monthlyImpact,
            isSimulated: true,
          ));
        } else if (result.monthlyImpact < 0) {
          monthExpense += result.monthlyImpact.abs();
          expenseItems.add(MonthLineItem(
            label: change.label,
            amount: result.monthlyImpact.abs(),
            isSimulated: true,
          ));
        }
      }

      final net = monthIncome - monthExpense;
      cumulative += net;

      projections.add(MonthProjection(
        yearMonth: ym,
        monthLabel: monthLabel,
        income: monthIncome,
        expense: monthExpense,
        net: net,
        cumulativeNet: cumulative,
        incomeItems: incomeItems,
        expenseItems: expenseItems,
      ));
    }

    return projections;
  }

  // ─── Amortization Schedule ───────────────────────────────────────

  static List<AmortizationRow> _amortizationSchedule({
    required double principal,
    required double annualRate,
    required int termMonths,
  }) {
    final r = annualRate / 12;
    final monthly = FinancialCalculator.monthlyLoanPayment(
      principal: principal,
      annualRate: annualRate,
      termMonths: termMonths,
    );
    double balance = principal;
    return List.generate(termMonths, (i) {
      final interest = balance * r;
      final principalPaid = monthly - interest;
      balance -= principalPaid;
      return AmortizationRow(
        month: i + 1,
        payment: monthly,
        principal: principalPaid,
        interest: interest,
        balance: balance.clamp(0, double.infinity),
      );
    });
  }

  // ─── Empty Result ────────────────────────────────────────────────

  static SimulationResult _emptyResult(
    MonthSummary budget,
    List<MonthLineItem> incomeItems,
    List<MonthLineItem> expenseItems,
    List<ProjectionBaseItem> baseItems,
  ) {
    final now = DateTime.now();
    return SimulationResult(
      currentIncome: budget.totalIncome,
      currentExpense: budget.totalExpense,
      currentNet: budget.netBalance,
      newIncome: budget.totalIncome,
      newExpense: budget.totalExpense,
      newNet: budget.netBalance,
      monthlyNetImpact: 0,
      annualNetImpact: 0,
      newSavingsRate: budget.savingsRate,
      newExpenseRate: budget.expenseRate,
      changeResults: [],
      monthlyProjection: List.generate(12, (i) {
        final m = DateTime(now.year, now.month + i);
        final ym = '${m.year}-${m.month.toString().padLeft(2, '0')}';

        double monthInc;
        double monthExp;
        List<MonthLineItem> mIncItems;
        List<MonthLineItem> mExpItems;

        if (baseItems.isNotEmpty) {
          monthInc = 0;
          monthExp = 0;
          mIncItems = [];
          mExpItems = [];
          for (final item in baseItems) {
            if (!item.isActiveOn(m)) continue;
            final amount = item.grossAmount != null
                ? FinancialCalculator.resolveNetForMonth(
                    amount: item.grossAmount!,
                    isGross: true,
                    month: m.month,
                  )
                : item.netAmount;
            if (item.isIncome) {
              monthInc += amount;
              mIncItems.add(MonthLineItem(label: item.label, amount: amount));
            } else {
              monthExp += amount;
              mExpItems.add(MonthLineItem(label: item.label, amount: amount));
            }
          }
        } else {
          monthInc = budget.totalIncome;
          monthExp = budget.totalExpense;
          mIncItems = incomeItems;
          mExpItems = expenseItems;
        }

        final net = monthInc - monthExp;
        return MonthProjection(
          yearMonth: ym,
          monthLabel: FinancialCalculator.monthShortNamesTR[(m.month - 1) % 12],
          income: monthInc,
          expense: monthExp,
          net: net,
          cumulativeNet: net * (i + 1),
          incomeItems: mIncItems,
          expenseItems: mExpItems,
        );
      }),
    );
  }

}

File: SavvyApp/Packages/Core/SavvyFoundation/Sources/SavvyFoundation/Calculator/SimulationCalculator.swift
import Foundation

public enum SimulationCalculator {

    // MARK: - Main Entry Point

    public static func calculateScenario(
        changes: [SimulationChange],
        currentBudget: MonthSummary,
        existingIncomeItems: [MonthLineItem] = [],
        existingExpenseItems: [MonthLineItem] = []
    ) -> SimulationResult {
        if changes.isEmpty {
            return emptyResult(currentBudget, existingIncomeItems, existingExpenseItems)
        }

        var totalIncomeImpact: Decimal = 0
        var totalExpenseImpact: Decimal = 0
        var changeResults: [ChangeResult] = []

        for change in changes {
            let result = calculateChange(change, budget: currentBudget)
            changeResults.append(result)
            if result.monthlyImpact > 0 {
                totalIncomeImpact += result.monthlyImpact
            } else {
                totalExpenseImpact += abs(result.monthlyImpact)
            }
        }

        let newIncome = currentBudget.totalIncome + totalIncomeImpact
        let newExpense = currentBudget.totalExpense + totalExpenseImpact
        let newNet = newIncome - newExpense
        let monthlyNetImpact = totalIncomeImpact - totalExpenseImpact

        let newExpenseRate = newIncome > 0
            ? NSDecimalNumber(decimal: newExpense / newIncome).doubleValue
            : 0.0
        let newSavingsRate = newIncome > 0
            ? min(max(NSDecimalNumber(decimal: (newIncome - newExpense) / newIncome).doubleValue, 0), 1)
            : 0.0

        var affordability: AffordabilityStatus? = nil
        if totalExpenseImpact > 0 {
            affordability = FinancialCalculator.loanAffordability(
                monthlyPayment: totalExpenseImpact,
                monthlyIncome: newIncome
            )
        }

        let projection = buildProjection(
            changes: changes,
            changeResults: changeResults,
            currentBudget: currentBudget,
            existingIncomeItems: existingIncomeItems,
            existingExpenseItems: existingExpenseItems
        )

        return SimulationResult(
            currentIncome: currentBudget.totalIncome,
            currentExpense: currentBudget.totalExpense,
            currentNet: currentBudget.netBalance,
            newIncome: newIncome,
            newExpense: newExpense,
            newNet: newNet,
            monthlyNetImpact: monthlyNetImpact,
            annualNetImpact: calculateAnnualImpact(changes, changeResults),
            newSavingsRate: newSavingsRate,
            newExpenseRate: newExpenseRate,
            affordability: affordability,
            changeResults: changeResults,
            monthlyProjection: projection
        )
    }

    // MARK: - Per-Change Calculators

    private static func calculateChange(_ change: SimulationChange, budget: MonthSummary) -> ChangeResult {
        switch change {
        case .credit(let principal, let annualRate, let termMonths, _):
            return calcLoanBased(
                change: change,
                loanPrincipal: principal,
                annualRate: annualRate,
                termMonths: termMonths,
                downPayment: 0,
                monthlyExtras: 0
            )
        case .housing(let price, let downPayment, let annualRate, let termMonths, let monthlyExtras, _):
            return calcLoanBased(
                change: change,
                loanPrincipal: price - downPayment,
                annualRate: annualRate,
                termMonths: termMonths,
                downPayment: downPayment,
                monthlyExtras: monthlyExtras
            )
        case .car(let price, let downPayment, let annualRate, let termMonths, let monthlyRunningCosts, _):
            return calcLoanBased(
                change: change,
                loanPrincipal: price - downPayment,
                annualRate: annualRate,
                termMonths: termMonths,
                downPayment: downPayment,
                monthlyExtras: monthlyRunningCosts
            )
        case .rentChange(let currentRent, let newRent, _, _):
            let diff = newRent - currentRent
            return ChangeResult(change: change, monthlyImpact: -diff)
        case .salaryChange(let currentGross, let newGross, _):
            return calcSalaryChange(change: change, currentGross: currentGross, newGross: newGross)
        case .income(let amount, _, _, _):
            return ChangeResult(change: change, monthlyImpact: amount)
        case .expense(let amount, _, _, _):
            return ChangeResult(change: change, monthlyImpact: -amount)
        case .investment(let principal, let annualReturnRate, let termMonths, let isCompound, _):
            return calcInvestment(change: change, principal: principal, annualReturnRate: annualReturnRate, termMonths: termMonths, isCompound: isCompound)
        }
    }

    private static func calcLoanBased(
        change: SimulationChange,
        loanPrincipal: Decimal,
        annualRate: Decimal,
        termMonths: Int,
        downPayment: Decimal,
        monthlyExtras: Decimal
    ) -> ChangeResult {
        guard loanPrincipal > 0 else {
            return ChangeResult(change: change, monthlyImpact: -monthlyExtras, totalCost: downPayment)
        }

        let rate = annualRate / 100
        let monthly = FinancialCalculator.monthlyLoanPayment(
            principal: loanPrincipal, annualRate: rate, termMonths: termMonths
        )
        let total = monthly * Decimal(termMonths)
        let interest = total - loanPrincipal
        let totalMonthly = monthly + monthlyExtras

        return ChangeResult(
            change: change,
            monthlyImpact: -totalMonthly,
            totalCost: total + downPayment + (monthlyExtras * Decimal(termMonths)),
            totalInterest: interest,
            amortizationSchedule: amortizationSchedule(
                principal: loanPrincipal, annualRate: rate, termMonths: termMonths
            )
        )
    }

    private static func calcSalaryChange(
        change: SimulationChange,
        currentGross: Decimal,
        newGross: Decimal
    ) -> ChangeResult {
        let currentNets = FinancialCalculator.resolveAllMonths(amount: currentGross, isGross: true)
        let newNets = FinancialCalculator.resolveAllMonths(amount: newGross, isGross: true)

        let currentAvg = currentNets.reduce(Decimal(0), +) / 12
        let newAvg = newNets.reduce(Decimal(0), +) / 12
        let monthlyDelta = newAvg - currentAvg

        return ChangeResult(
            change: change,
            monthlyImpact: monthlyDelta,
            salaryImpact: SalaryImpact(oldNet: currentAvg, newNet: newAvg)
        )
    }

    private static func calcInvestment(
        change: SimulationChange,
        principal: Decimal,
        annualReturnRate: Decimal,
        termMonths: Int,
        isCompound: Bool
    ) -> ChangeResult {
        let totalReturn: Decimal
        let p = NSDecimalNumber(decimal: principal).doubleValue
        let rate = NSDecimalNumber(decimal: annualReturnRate / 100).doubleValue

        if isCompound {
            let monthlyRate = rate / 12
            let result = p * (pow(1 + monthlyRate, Double(termMonths)) - 1)
            totalReturn = Decimal(result)
        } else {
            let result = p * rate * (Double(termMonths) / 12)
            totalReturn = Decimal(result)
        }

        let monthlyReturn = totalReturn / Decimal(termMonths)
        return ChangeResult(
            change: change,
            monthlyImpact: monthlyReturn,
            investmentImpact: InvestmentImpact(
                totalReturn: principal + totalReturn,
                totalInvested: principal
            )
        )
    }

    // MARK: - Annual Impact

    private static func calculateAnnualImpact(
        _ changes: [SimulationChange],
        _ changeResults: [ChangeResult]
    ) -> Decimal {
        var annual: Decimal = 0
        for i in 0..<changes.count {
            let change = changes[i]
            let result = changeResults[i]

            if case .salaryChange = change, let si = result.salaryImpact {
                annual += (si.newNet - si.oldNet) * 12
            } else {
                let effectiveMonths: Int
                switch change {
                case .credit(_, _, let term, _),
                     .housing(_, _, _, let term, _, _),
                     .car(_, _, _, let term, _, _),
                     .investment(_, _, let term, _, _):
                    effectiveMonths = min(term, 12)
                default:
                    effectiveMonths = 12
                }
                annual += result.monthlyImpact * Decimal(effectiveMonths)
            }
        }
        return annual
    }

    // MARK: - 12-Month Projection

    private static func buildProjection(
        changes: [SimulationChange],
        changeResults: [ChangeResult],
        currentBudget: MonthSummary,
        existingIncomeItems: [MonthLineItem],
        existingExpenseItems: [MonthLineItem]
    ) -> [MonthProjection] {
        let now = Date()
        let calendar = Calendar.current
        var projections: [MonthProjection] = []
        var cumulative: Decimal = 0

        for i in 0..<12 {
            guard let projDate = calendar.date(byAdding: .month, value: i, to: now) else { continue }
            let ym = projDate.toYearMonth()
            let monthIdx = (calendar.component(.month, from: projDate) - 1) % 12
            let monthLabel = FinancialCalculator.monthShortNamesTR[monthIdx]

            var monthIncome = currentBudget.totalIncome
            var monthExpense = currentBudget.totalExpense
            var incomeItems = existingIncomeItems
            var expenseItems = existingExpenseItems

            for ci in 0..<changes.count {
                let change = changes[ci]
                let result = changeResults[ci]

                let isWithinTerm: Bool
                switch change {
                case .credit(_, _, let term, _),
                     .housing(_, _, _, let term, _, _),
                     .car(_, _, _, let term, _, _),
                     .investment(_, _, let term, _, _):
                    isWithinTerm = i < term
                default:
                    isWithinTerm = true
                }
                guard isWithinTerm else { continue }

                // Rent with annual increase
                if case .rentChange(let currentRent, let newRent, let annualIncrease, let label) = change,
                   annualIncrease > 0 {
                    let baseDiff = newRent - currentRent
                    let yearsPassed = i / 12
                    let rate = NSDecimalNumber(decimal: annualIncrease / 100).doubleValue
                    let adjustedDiff = NSDecimalNumber(decimal: baseDiff).doubleValue * pow(1 + rate, Double(yearsPassed))
                    let decimalDiff = Decimal(adjustedDiff)
                    monthExpense += decimalDiff
                    expenseItems.append(MonthLineItem(label: label, amount: decimalDiff, isSimulated: true))
                    continue
                }

                // Salary change per-month
                if case .salaryChange(let currentGross, let newGross, let label) = change {
                    let currentNets = FinancialCalculator.resolveAllMonths(amount: currentGross, isGross: true)
                    let newNets = FinancialCalculator.resolveAllMonths(amount: newGross, isGross: true)
                    let delta = newNets[monthIdx] - currentNets[monthIdx]
                    monthIncome += delta
                    incomeItems.append(MonthLineItem(label: label, amount: delta, isSimulated: true))
                    continue
                }

                if result.monthlyImpact > 0 {
                    monthIncome += result.monthlyImpact
                    incomeItems.append(MonthLineItem(label: change.displayLabel, amount: result.monthlyImpact, isSimulated: true))
                } else if result.monthlyImpact < 0 {
                    monthExpense += abs(result.monthlyImpact)
                    expenseItems.append(MonthLineItem(label: change.displayLabel, amount: abs(result.monthlyImpact), isSimulated: true))
                }
            }

            let net = monthIncome - monthExpense
            cumulative += net

            projections.append(MonthProjection(
                yearMonth: ym, monthLabel: monthLabel,
                income: monthIncome, expense: monthExpense,
                net: net, cumulativeNet: cumulative,
                incomeItems: incomeItems, expenseItems: expenseItems
            ))
        }
        return projections
    }

    // MARK: - Amortization Schedule

    private static func amortizationSchedule(
        principal: Decimal, annualRate: Decimal, termMonths: Int
    ) -> [AmortizationRow] {
        let r = NSDecimalNumber(decimal: annualRate / 12).doubleValue
        let monthly = FinancialCalculator.monthlyLoanPayment(
            principal: principal, annualRate: annualRate, termMonths: termMonths
        )
        let monthlyD = NSDecimalNumber(decimal: monthly).doubleValue
        var balance = NSDecimalNumber(decimal: principal).doubleValue

        return (0..<termMonths).map { i in
            let interest = balance * r
            let principalPaid = monthlyD - interest
            balance = max(balance - principalPaid, 0)
            return AmortizationRow(
                month: i + 1,
                payment: monthly,
                principal: Decimal(principalPaid),
                interest: Decimal(interest),
                balance: Decimal(balance)
            )
        }
    }

    // MARK: - Empty Result

    private static func emptyResult(
        _ budget: MonthSummary,
        _ incomeItems: [MonthLineItem],
        _ expenseItems: [MonthLineItem]
    ) -> SimulationResult {
        let now = Date()
        let calendar = Calendar.current
        let projection = (0..<12).map { i -> MonthProjection in
            let projDate = calendar.date(byAdding: .month, value: i, to: now)!
            let ym = projDate.toYearMonth()
            let monthIdx = (calendar.component(.month, from: projDate) - 1) % 12
            let net = budget.totalIncome - budget.totalExpense
            return MonthProjection(
                yearMonth: ym,
                monthLabel: FinancialCalculator.monthShortNamesTR[monthIdx],
                income: budget.totalIncome,
                expense: budget.totalExpense,
                net: net,
                cumulativeNet: net * Decimal(i + 1),
                incomeItems: incomeItems,
                expenseItems: expenseItems
            )
        }

        return SimulationResult(
            currentIncome: budget.totalIncome,
            currentExpense: budget.totalExpense,
            currentNet: budget.netBalance,
            newIncome: budget.totalIncome,
            newExpense: budget.totalExpense,
            newNet: budget.netBalance,
            monthlyNetImpact: 0,
            annualNetImpact: 0,
            newSavingsRate: budget.savingsRate,
            newExpenseRate: budget.expenseRate,
            affordability: nil,
            changeResults: [],
            monthlyProjection: projection
        )
    }
}
